### ISSUE TITLE
Python version message gets into comint's history.

### ISSUE THREAD

**Issue by gety9** [jdtsmith/python-mls#10]
From time to time "python version message" gets into history (usually when i cycle history back), example:

```
>>> Python 3.11.1 (main, Jan 16 2023, 08:53:45) [Clang 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0 on linux
... Type "help", "copyright", "credits" or "license" for more information.
... >>> 

```

and when i try to delete the above, it says 

`text is read-only`

so i have to S-RET, cause error, and then clean all comint.

---

One more time thank you for great package!

**Comment by jdtsmith** [comment]
OK thanks; now we need to reproduce this.  Please let me know your system, Emacs version, etc.  

Does this even occur right away on loading Python?  Or only after some time?  See if you can find a way to reproduce.  First you need to remove history files to get that errant entry out; you can get rid of ~/.emacs.d/pyhist-* (or just the relevant one).

**Comment by gety9** [comment]
`GNU Emacs 28.2 (build 1, aarch64-unknown-linux-android, X toolkit, Xaw scroll bars) of 2023-01-17`

```
termux-tools version:
1.35.0
Android version:
11
Kernel build information:
Linux localhost 4.19.157-perf-g616268c8715f-dirty #134 SMP PREEMPT Wed Nov 10 15:13:53 CST 2021 aarch64 Android
```

"First you need to remove history files to get that errant entry out; you can get rid of ~/.emacs.d/pyhist-*"

-, will try do now and see if i can reproduce.

**Comment by gety9** [comment]
@jdtsmith no luck yet :) maybe i did smng out of ordinary that time to get python version message in pyhist :)

By the way could you tell what's the best way to make mls always enabled when i use M-x run python?

Cause now first time during session i always have to manually M-x python-mls-mode... 

(i don't think that attaching to python-mode hook is way to do it cause M-x run-python is not it, right?)

**Comment by jdtsmith** [comment]
I have seen it happen myself, so I don't think it's just you.  Before you start, you might `M-x trace-function Ret comint-add-to-input-history`.  This command is where input history actually gets saved (by comint), so you'll get a report when this is run.  Keep an eye on the `*trace-output*` produced, and then try doing various things when you run-python for the first time, to get it to stick the intro text into the history. 

> what's the best way to make mls always enabled

Did you check Installation notes?

**Comment by gety9** [comment]
@jdtsmith do you mean

```
(use-package python-mls
  :config
  (python-mls-setup))
```

?

yes, i have this part:

```
(use-package python-mls
  :ensure t
  :bind (:map python-mls-mode-map
	      ;; ("<up>" . python-mls-up-or-history)
	      ;; ("<down>" . python-mls-down-or-history)
	      ;; ("M-<up>" . comint-previous-matching-input-from-input)
              ;; ("M-<down>" . comint-next-matching-input-from-input)
	      ;; ("S-<up>" . python-mls-noblock-up-or-history)
	      ;; ("S-<down>" . python-mls-noblock-down-or-history)
	      ("<return>" . newline-and-indent)
	      ("S-<delete>" . comint-clear-buffer))
  :config
  (python-mls-setup))
```

and it's still not auto enabling, so i assumed that i missed smng.

**Comment by jdtsmith** [comment]
> it's still not auto enabling

That's strange.  Check the variable `inferior-python-mode-hook`.  It should contain `python-mls-mode`. 

Update: I see you have rebound `<return>`; you can't do that.  python-mls checks for continuation first on return.  This binding defeats that.  It also already indents for you.

**Comment by gety9** [comment]
nil

```
inferior-python-mode-hook is a variable defined in 'python.el'.

Its value is nil
```

**Comment by jdtsmith** [comment]
`(python-mls-setup)` definitely sets this, so something is wrong with your config here. Start over in a clean session and report.

**Comment by gety9** [comment]
@jdtsmith is there smng that can prevernt python-mls-setup from running or undo inferior-python-mode-hook?

Asking cause if i manually do M-x python-mls-setup it works fine (inferior-python-mode-hook has python-mls-mode) for 1 session (untell emacs kill)

**Comment by jdtsmith** [comment]
You are supposed to run `(python-mls-setup)` once in your config, as per Installation instructions.  Then not again.  If you do that and your config is functional, the hook should be setup.  

> is there smng that can prevernt python-mls-setup from running or undo inferior-python-mode-hook?

Of course: any configuration issue (typo, missing package, whatever) can cause parts of your init to not run.  Check `*Messages*`. This is I am quite sure a config issue on your end; if you disagree, please isolate the issue via `emacs -Q` and open another issue. 

For the current issue please see if you can reproduce the intro text getting placed on history.

**Comment by gety9** [comment]
@jdtsmith 

thank you for explanation, agree python-mls-setup issue is likely on my side...

Good news :) i seem to found a way to reliably (100%) put python intro message in comint history, steps:

```
1, M-x run-python
2, S-down
3, S-up
```

**Comment by jdtsmith** [comment]
Hmm, that doesn't seem to do it for me.  Does your down-up trick work if you start with a clean history, deleting the pyhist file and restarting emacs?  Are you doing it while the initial text arrives, or after?  Once the junk is in the history, these commands will recall it, so you may think they are being added.  That's why you must delete the appropriate pyhist-* file.

**Comment by jdtsmith** [comment]
BTW, you must delete ~/emacs.d/pyhist-Python _after_ you exit the interpreter and kill that buffer, since by default it writes its history when this happens.

**Comment by gety9** [comment]
@jdtsmith 

1. "Does your down-up trick work if you start with a clean history, deleting the pyhist file and restarting emacs?"

Exact steps:
```
(clean history)
1 Kill python enterpreter
2 Kill emacs
3 Delete pyhist-python with bash or whatever
(enter 1 harmless record in history)
4 Start emacs
5 M-x run python
6 print('A') + S-Ret
7 Kill python enterpeter
7 close buffer
8 Kill emacs
(bug)
9 Start emacs
10 M-x run python
11 S-down
>>> print('A')
12 S-up
>>> Python 3.11.1 (main, Jan 16  2023......
```

**Comment by gety9** [comment]
2. I fixed my python-mls-setup issue by changing `config` to `init`

```
(use-package python-mls
  :ensure t
  :bind (:map python-mls-mode-map
	      ;; ("<up>" . python-mls-up-or-history)
	      ;; ("<down>" . python-mls-down-or-history)
	      ;; ("M-<up>" . comint-previous-matching-input-from-input)
              ;; ("M-<down>" . comint-next-matching-input-from-input)
	      ;; ("S-<up>" . python-mls-noblock-up-or-history)
	      ;; ("S-<down>" . python-mls-noblock-down-or-history)
	      ("<return>" . newline-and-indent)
	      ("S-<delete>" . comint-clear-buffer))
  :init
  (python-mls-setup))
```

It seems like (python-mls-setup) works only if it's not deferred by use-package, and use-package defers `config` part if there is `bind`, consider using `init` in docs (since pyton-mls-setup command seems to require not deferring)

**Comment by jdtsmith** [comment]
Thanks for this very helpful reproduction recipe.  The key is to put something in the history ring, then attempt to recall input from history and "return" then to the command line. 

You have revealed a bug in python.el's setup code, which results in the first prompt not being marked "field=output".   You can see this manifested as well in `comint-bol` (`C-c C-a`) going before the first prompt (only).  

This incorrect labeling causes the history mechanism's recall of the "saved input" to mistakenly identify the prompt (and possibly material before it) as _input_.  Once this erroneous "input" gets copied to the current prompt, since it's marked read-only, you can't get rid of it, so naturally hit Return, putting it on the history.  I have just uploaded a workaround fix; please give it a try and let me know how that goes.

**Comment by jdtsmith** [comment]
Thanks as well for your feedback on the use-package.  I've recommended `:after python` instead, which should also work.  If you could try that (with the `:config`) and let me know, I'd appreciate.

**Comment by gety9** [comment]
@jdtsmith per you suggestion tried :after python, and it didn't work

as i understand :init and :after python achieve 2 different things:
-, :after python  + :config runs config after python is loaded, but it can still be run after python-mls is loaded (if :bind is used)
-, :init makes sure config is run before python-mls is loaded

https://jwiegley.github.io/use-package/keywords/
"Use the :init keyword to execute code before a package is loaded. It accepts one or more forms, up until the next keyword:"

**Comment by gety9** [comment]
Basically this didn't work

```
(use-package python-mls
  :ensure t
  :after python
  :bind (:map python-mls-mode-map
	      ("<up>" . previous-line)
	      ("<down>" . next-line)
	      ("C-<up>" . backward-paragraph)
	      ("C-<down>" . forward-paragraph)
	      ;; ("M-<up>" . comint-previous-matching-input-from-input)
              ;; ("M-<down>" . comint-next-matching-input-from-input)
	      ;; ("S-<up>" . python-mls-noblock-up-or-history)
	      ;; ("S-<down>" . python-mls-noblock-down-or-history)
	      ("<return>" . newline-and-indent)
	      ("S-<delete>" . comint-clear-buffer))
  :config
  (python-mls-setup))
```

i think cause of :bind (which also imply defer)

"In almost all cases you don’t need to manually specify :defer t. This is implied whenever :bind or :mode or :interpreter is used."

**Comment by gety9** [comment]
(python version intro message in history bug)

"I have just uploaded a workaround fix; please give it a try and let me know how that goes."

-, it fixed it, thank you!

**Comment by jdtsmith** [comment]
For your load issue: here's what needs to happen:

1. python.el is loaded.
2. python-mls is then _immediately_ loaded.
3. `(python-mls-setup)` then gets called.

So yes, with your use of `:bind`, step 2 is setup to occur only after the bound commands get loaded (which in this case isn't appropriate). So either add `:defer nil`, or do your binding after `python-mls` is loaded, i.e. in the `:config` stanza.  

By the way, you can `M-x pp-macroexpand-last-sexp` at the end of a use-package stanza and see all of the details.

**Comment by jdtsmith** [comment]
> Use the :init keyword to execute code before a package is loaded

In this case, what you are trying to execute in `:init` will _cause_ python-mls to get loaded (via an autoload).  `:init` is good for defining useful functions or other things you need to have in place before the mode gets loaded.

**Comment by jdtsmith** [comment]
For posterity, the report for the python.el bug which causes intro/prompt text to get misinterpreted as incomplete input [is here](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=61523).

**Comment by jdtsmith** [comment]
I just simplified the setup for use-package, eliminating `python-mls-setup`.  This should make `:bind` and all that less error-prone. Please take a look and let me know of any issue.

**Comment by jdtsmith** [comment]
Did you have a chance to take a look at the new config method?  It does require a change to your config.
